### PR TITLE
Support `tags` parameter in illust_bookmark_add(), improve JsonDict attr accessing

### PR DESCRIPTION
**Pull request by invobzvr**
Maybe easier to check auth
```py
if res.error and 'invalid_grant' in res.error.message:
    self.auth()
```

### CODE REVIEW

**Comment by upbit** [comment]
Thanks for the contribution!

Could you add some example for `unquote tags` in demo.py, in order to test API is working?

**Comment by invobzvr** [comment]
Sorry for a missing change, it should be `data['tag'] = quote(tags)`
And the API just worked fine hours ago, but didn't work just now ... and sorry for that  i haven't found the reason yet, maybe you can have a try on your devices

File: pixivpy3/api.py [hunk 1]
```diff
@@ -42,15 +42,7 @@ def set_accept_language(self, language):
...
-
-        def _obj_hook(pairs):
-            """convert json object to python object"""
-            o = JsonDict()
-            for k, v in pairs.items():
-                o[str(k)] = v
-            return o
-
-        return json.loads(json_str, object_hook=_obj_hook)
+        return json.loads(json_str, object_hook=lambda obj: JsonDict(obj))
```
Comment (Xdynix) [inline pixivpy3/api.py:45]: It could be simplified to `return json.loads(json_str, object_hook=JsonDict)`

File: pixivpy3/aapi.py [hunk 2]
```diff
@@ -58,18 +70,12 @@ def format_bool(self, bool_value):
...
-        else:
-            from urlparse import urlparse, unquote
-            safe_unquote = lambda s: unquote(s.encode('utf8')).decode('utf8')
 
         result_qs = {}
         query = urlparse(next_url).query
         for kv in query.split('&'):
             # split than unquote() to k,v strings
-            k, v = map(safe_unquote, kv.split('='))
+            k, v = map(unquote, kv.split('='))
```
Comment (Xdynix) [inline pixivpy3/aapi.py:78]: You can use `urllib.parse.parse_qs` instead of unquote it by yourself.

File: pixivpy3/api.py (hunk 1 above)
Comment (invobzvr) [reply pixivpy3/api.py]: Thanks for reminding
First time to use object_hook

File: pixivpy3/aapi.py (hunk 2 above)
Comment (invobzvr) [reply pixivpy3/aapi.py]: Good job, i totally forgot that lib and didn't notice what it implemented

File: pixivpy3/aapi.py [hunk 3]
```diff
@@ -58,28 +63,12 @@ def format_bool(self, bool_value):
...
-                # keep the origin sequence, just ignore group('idx')
-                result_qs[mk] = marray + [v]
-            else:
-                result_qs[k] = v
+
+        query = up.urlparse(next_url).query
+        result_qs = up.parse_qs(query)
+        for key in result_qs:
+            if len(result_qs[key]) == 1:
+                result_qs[key] = result_qs[key][0]
```
Comment (upbit) [inline pixivpy3/aapi.py:71]: The returned params contains an array like `seed_illust_ids[]`, which needs to be parsed into a python array

Comment (upbit) [reply pixivpy3/aapi.py:71]: See [line 75](https://github.com/upbit/pixivpy/pull/124/files/f862895dde6e0f2ac991f6c6d56bb29910e4853c#diff-f603bba12ba259bd95cdcb4f29726d3fL75)

Comment (invobzvr) [reply pixivpy3/aapi.py]: Sorry for my negligence

**Comment by upbit** [comment]
Sorry for late response, merged.

**Comment by upbit** [comment]
There is a python2 character set problem, it will use UTF-8 instead of unicode by default:

```py
# python3.7: search_illust() -> parse_qs(next_url), word is unicode
{'sort': 'date_desc', 'search_target': 'partial_match_for_tags', 'word': u'\u843d\u66f8\u304d', 'filter': 'for_ios'}

# python2.7: search_illust() -> parse_qs(next_url), word is UTF-8
{'sort': u'date_desc', 'search_target': u'partial_match_for_tags', 'word': u'\xe8\x90\xbd\xe6\x9b\xb8\xe3\x81\x8d', 'filter': u'for_ios'}
```

It will cause [encoder error](https://github.com/upbit/pixivpy/pull/125/checks?check_run_id=599065279) when get next page:

```py
https://210.140.131.226/v1/search/illust?sort=date_desc&search_target=partial_match_for_tags&word=%C3%A8%C2%90%C2%BD%C3%A6%C2%9B%C2%B8%C3%A3%C2%81%C2%8D&filter=for_ios&offset=30
{u'illusts': [], u'search_span_limit': 31536000, u'next_url': None}
```

**Comment by upbit** [comment]
fixed in #126

**Pull request merged**